Change a collection of trait functions into constants

SyxtonPrime opened the pull request:
(Replacing this heading comment by a later comment which better reflects this PR after some updates)

This PR looks a lot bigger than it really is as most of the changes are simply updating notation. (E.g. replacing `F::one()` by `F::ONE` throughout the code base). The main motivation for this change is "moral" as opposed to performance focused. This might though possible lead to some very small speedups if function inlining was not happening in some cases. Testing on the example `prove_poseidon2_koala_bear_keccak` the speed seems to be basically unchanged (at best there is maybe a `1-2%` speed up but this sort of tiny improvement is very hard to pick up)

The major changes are as follows:

## Changes to Field Traits

- In the AbstractField trait, the functions `zero()`, `one()`, `two()` and `neg_one()` have all been replaced by constants with the same name but capitalized.
- The function `generator()` has been removed from the AbstractField trait and replaced by a constant `GENERATOR` which can be found in the Field trait.
- In the BinomiallyExtendable trait, the functions  `w(), dth_root()` and `ext_generator()` have all been replaced by constants with the same name but capitalized. 
- In the ComplexExtendable trait, the function `complex_generator()` has all been replaced by a constant with the same name but capitalized. 
- In the HasComplexBinomialExtension trait, the functions  `w(), dth_root()` and `ext_generator()` have all been replaced by constants with the same name but capitalized.

## New Additions:
- A major rework to field_to_array in order to make it a constant function which is needed to generate constants for binomial extensions. This involves a somewhat sketchy workaround as none of the more natural methods are available in const environments on stable. If this merges I'll make an issue just to keep an eye on this code and get rid of it when possible. Happy to change the name and/or move it into utils or something if it fits better there.
- Added a couple of new tests to field-testing to check `F::ZERO, F::ONE, F::TWO, F::NEG_ONE` as some of these were previously untested.

dlubarov commented:
We did have it as a const originally, I had changed it in #160. I think the main complication is `BinomialExtensionField` - not sure if/how we can rewrite its other functions like `one()` with const code. (We can't do things like `[AF::zero(); D]` since that expects `Copy`.)

dlubarov commented:
But if there is some reasonable way to do this for the other functions like `one()`, I agree this would be an improvement - a const feels natural and should remove the need to force inlining.

SyxtonPrime commented:
Having a look into the code, it seems like what we'd need is a `const fn` version of `field_to_array`.

Super annoyingly it seems that RUST isn't currently able to support this on stable. (I think this would be possible in nightly using MaybeUninit's). I'll keep playing around with it to see if I get anywhere but this seems like a known rust issue.

SyxtonPrime commented:
One option (in the short term) would be to add the required constants into the BinomiallyExtendable trait and then copy them over.

It's a little inelegant but might be worth it?

SyxtonPrime commented:
Realized I needed `AF::ZERO` to be a constant to get my fix working so I've just reopened this PR.

It now lets us replace both `zero()` and `one()` but constant functions.

It is possible to do the same for `two()` and `neg_one()` but we need to first do a refactor on the bn-254 crate.

The main questions for this PR which it would be good it get input from a couple of people on are:
- What do we think of the "fix" to field_to_array to make it const? Is this too hacky and not worth having in the code base?
- `field_to_array` is also used in a couple of non-const places (in particular in implementing `From` and `from_base` for `BinomialExtensionField<AF, D>`. Should we have a second function which does the far simpler impl of `field_to_array` and works in the non-const environment?

E.g.: (Or potentially call this one `field_to_array` and the other one `const_field_to_array`)
```
fn non_const_field_to_array<AF: AbstractField, const D: usize>(x: AF) -> [AF; D]
      let mut arr: [AF; D] = [AF::ZERO; D];
      arr[0] = x
      arr
```

In principle the two functions should compile to the same thing but the rust compiler might do strange things.

SyxtonPrime commented:
This PR looks a lot bigger than it really is as most of the changes are simply updating notation. (E.g. replacing `F::one()` by `F::ONE` throughout the code base). The main motivation for this change is "moral" as opposed to performance focused. This might though possible lead to some very small speedups if function inlining was not happening in some cases. Testing on the example `prove_poseidon2_koala_bear_keccak` the speed seems to be basically unchanged (at best there is maybe a `1-2%` speed up but this sort of tiny improvement is very hard to pick up)

The major changes are as follows:

## Changes to Field Traits

- In the AbstractField trait, the functions `zero()`, `one()`, `two()` and `neg_one()` have all been replaced by constants with the same name but capitalized.
- The function `generator()` has been removed from the AbstractField trait and replaced by a constant `GENERATOR` which can be found in the Field trait.
- In the BinomiallyExtendable trait, the functions  `w(), dth_root()` and `ext_generator()` have all been replaced by constants with the same name but capitalized. 
- In the ComplexExtendable trait, the function `complex_generator()` has all been replaced by a constant with the same name but capitalized. 
- In the HasComplexBinomialExtension trait, the functions  `w(), dth_root()` and `ext_generator()` have all been replaced by constants with the same name but capitalized.

## New Additions:
- A major rework to field_to_array in order to make it a constant function which is needed to generate constants for binomial extensions. This involves a somewhat sketchy workaround as none of the more natural methods are available in const environments on stable. If this merges I'll make an issue just to keep an eye on this code and get rid of it when possible. Happy to change the name and/or move it into utils or something if it fits better there.
- Added a couple of new tests to field-testing to check `F::ZERO, F::ONE, F::TWO, F::NEG_ONE` as some of these were previously untested.

nbgl commented on field/src/helpers.rs:
```diff
@@ -69,19 +69,80 @@ where
+    elements: ManuallyDrop<[MaybeUninit<T>; D]>,
+}
+
+impl<T, const D: usize> HackyWorkAround<T, D> {
```
I hate this but yeah I'm convinced that it works and I couldn't find an easier way.
